Loan and Exchange management tools improvements

mickley opened the pull request:
This pull request fixes a few bugs/quirks and adds some features to the loan and exchange management tools. 

In brief, but see individual commit messages:
- Fixes bug we discovered at the last data portal lunch break where transactions (especially exchanges) become inaccessible/not editable when an institution does not have a code. This falls back to using the institution name. 
- Adds a due date to list of outgoing or incoming loans, if specified. This was a request for loan tracking by Dawn Neisser at the last lunch break.
- Improved autogeneration of the next transaction/loan identifier number; this had some bugs, and didn't work well for more complex identifiers
- Finally, this adds an attachment system to the loans and exchanges to allow for saving correspondence as attachments along with the loan.

mickley committed:
Adds due date to list of outgoing or incoming loans, if specified

This is helpful to know when loans are due. Furthermore, if the due date is earlier than the current date (on server time) then the due date will be highlighted in red as overdue.

If loans are closed, the due date is no longer shown.

mickley committed:
Fixes bug in exchange and loans lists.

Exchanges from institutions without a code become inaccessable, since there is no code to click on to expand the list of gifts/exchanges from that institution. Loans from institutions without a code are not inaccessible, but there is no way to tell where they are from without editing the.

This fixes both problems by preferring the code to show an institution, but falling back to the institution name if code is not specified.

mickley committed:
Adds a file attachment system to loans and exchanges

This allows a collection editor to save digital files with the correspondence attached to a particular shipment: shipping notices, loan acknowledgements, specimen lists, signed agreements, emails, and any other information relevant to that loan, exchange, or gift. Optionally, users can provide a descriptive title for each attachment.

The maximum attachment size is set at 10 MB, and files are currently restricted to the following filetypes: .pdf, .jpg, .jpeg, .png, .txt, .csv, .xls, .xlsx, .doc, and .docx.

Files are physically saved in content/collections/loans under a subfolder with the collection ID. The server needs to have write access to content/collections/loans. To minimize duplicate filenames, the loan or transaction number and the date are appended to the original filename. Files are also logged in a new database table omoccurloansattachments, the schema of which is in config/schema/dev/db_schema_patch-dev.sql.

mickley committed:
Improved autogeneration of next loan/exchange ID

The previous code took the last three loans/exchanges, pulled the number or numbers out of the identifier, and returned the largest number found, incremented by one. However, there was a bug that prevented numbers in some identifiers from being counted at all, by skipping any identifier that had non-numeric characters in front of a number. In addition, taking the highest number found within an identifier with multiple numbers might be incorrect (e.g., for `2022-10` the code would increment 2022, not 10). Finally, any alphanumeric characters would not be included in the new identifier.

This fix solves all of these problems
- It returns a full alphanumeric identifier with the number incremented by one, if the largest recent identifier is alphanumeric
- It uses the *last* number found in a string, which is most likely the one to increment
- It should not skip some identifier formats

mickley commented:
Here's a screenshot of the loan due dates. Overdue loans are highlighted in red:
![loan-due-date](https://user-images.githubusercontent.com/9954229/178634372-b01fa637-9519-49b9-a869-23a23a1ee750.png)

mickley commented:
And, here's a screenshot of the correspondence attachment system for loans and exchanges:
![loan-attachments](https://user-images.githubusercontent.com/9954229/178634501-c58ef4b2-1b17-405c-abf2-7e05650559c0.png)

 This is a more complex change and involves a new database table (`omoccurloansattachments`) and saving files to content/collections/loans, which needs server write permissions. 

There are a number of digital records that would be useful to attach to a particular loan or exchange: shipping notices, loan acknowledgements, specimen lists, signed agreements, emails, etc.. I've tried to anticipate the file types that these would use, but there are probably more. 

As coded, the maximum attachment size is set at 10 MB, and files are currently restricted to the following filetypes: .pdf, .jpg, .jpeg, .png, .txt, .csv, .xls, .xlsx, .doc, and .docx.

GregoryPost reviewed (commented):
Would uploaded loan documents contain any confidential / sensitive information?  If so we may need to come up with a more secure way of storing the files on the server.  A web server misconfiguration could allow content/collections/loans to be browsed anonymously.  A potential attacker could also have some success at guessing filenames (challenging, but not impossible given the source code is open source.  ;-)  )

mickley commented:
@GregPost-ASU: Hmm, good question. 

Typical loan correspondence would not have any sensitive information. It would generally be limited to addresses that are already on portals, GBIF, etc. and lists of specimens, not including localities. 

At any rate, here's an example of what is typical:
[WTU Loan Return.pdf](https://github.com/BioKIC/Symbiota/files/9105767/WTU.Loan.Return.pdf)

A typical shipment would generate 2-4 of these, depending on whether a collection saved signed acknowledgements, or unsigned copies or both. 

It's conceivable that a collection manager could attach something sensitive but this would be atypical I think. I can't think of much beyond the contents of emails, or unpublished data of some sort that could fall into this category. Most emails that I would save would be an elaboration on what is requested, what the specimens would be used for, loan terms, destructive sampling, etc. 

Guessing filenames would be difficult, but possible I suppose. One would need to guess the date, the institution's transaction numbering scheme (which varies), as well as the original filename of the attachment (which is kept, and not in the source code). Loan numbers aren't publicly visible through Symbiota, so without some knowledge of an institution's numbering and predictable attachment naming schemes, it would be tough without a brute force attack.

mickley commented:
I should also say that saving this correspondence is very helpful, and I often go back to loan correspondence dating back 10-20 years. For me currently, that's rummaging through a filing cabinet. 

Sometimes there are handwritten notes on the paperwork where specimen counts were wrong, and these don't always get captured in Symbiota. Sometimes, there is no list of individual specimens with the paperwork itself, but a thoughtfully saved email has this, and is helpful to cross reference when the loan gets returned. Or, an incoming loan has a specimen list that could get lost. Or, several years later, an institution we sent something to claims they never got the package, so it is helpful to see if they acknowledged receipt.

mickley committed:
Added some additional security checks to file uploaded

- Confirm that it's an uploaded file
- Make sure the filename is not too long
- Check mimetype, don't rely on extension

mickley committed:
Fixes syntax error

mickley committed:
Fixes spelling error

egbot committed:
Merge branch 'master' into loans-improvements
